feat(chores): Add missing tools to nix-shell

mhkarimi1383 opened the pull request:
## Pull Request template

Please, go through these steps before you submit a PR.

**Why is this PR required? What issue does it fix?**: controller-tools and some other tools are missing from `nix-shell`

**What this PR does?**: adds corresponding packages

**Does this PR require any upgrade changes?**: No

**If the changes in this PR are manually verified, list down the scenarios covered:**:

just run `nix-shell`

**Any additional information for your reviewer?** :
_Mention if this PR is part of any design or a continuation of previous PRs_


**Checklist:**
- [ ] Fixes #<issue number>
- [x] PR Title follows the convention of  `<type>(<scope>): <subject>`
- [ ] Has the change log section been updated?
- [ ] Commit has unit tests
- [ ] Commit has integration tests
- [x] (Optional) Are upgrade changes included in this PR? If not, mention the issue/PR to track:
- [x] (Optional) If documentation changes are required, which issue on https://github.com/openebs/openebs-docs is used to track them:

mhkarimi1383 committed:
:technologist: Add `kubernetes-controller-tools` to nix-shell

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 commented:
I also want to add .envrc to automatically load that

tiagolobocastro commented:
Pardon my ignorance, what's in `kubernetes-controller-tools` ?

mhkarimi1383 commented:
@tiagolobocastro 
controller-gen cli

dsharma-dc commented:
Though nothing harmful, do we really need this in product nix shell ? The required controller-gen is currently installed from the Makefile in the repo and is used to generate manifests.

mhkarimi1383 commented:
In my NixOS I ran `make -j 1` and controller-gen not found and with the latest one from nix repos (unstable) I got unrecognized option error for controller-gen command

tiagolobocastro commented:
> Though nothing harmful, do we really need this in product nix shell ? The required controller-gen is currently installed from the Makefile in the repo and is used to generate manifests.

Downloaded binaries are not likely to work on nix and NixOS, may require patching et all.
It's simpler to just put it on the nix-shell rather than figure out why it doesn't work.

mhkarimi1383 commented:
I made this PR draft since I see that some more tools are missing....

I want to add every required tools to shell.nix to make easier to work for NixOS users :)

mhkarimi1383 committed:
:technologist: Add required Go version to nix shell

NOTE: after upgrading golang version we have to update this file,
also updating nixpkgs will break this since go1.19 is not available to
newer versions of nixpkgs

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:technologist: Add kubernetes tools to shell.nix

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:technologist: Add minikube to shell.nix

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:memo: Add nix shell to developer setup docs

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:technologist: Adding more external required tools to `shell.nix`

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:coffin: Remove minikube package option

Our nixpkgs is a bit old

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 commented:
Commit notes/descriptions are important

mhkarimi1383 committed:
:technologist: Add make to nix shell

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 commented:
Also we can change all of the actions to use nix shell, so changing go version, etc., will be easier

mhkarimi1383 committed:
:bug: `make` nix package name

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

tiagolobocastro commented:
Sounds good, just let us know when this is working and ready to merge 👍

mhkarimi1383 commented:
@tiagolobocastro 
Looks like ready, But since I'm getting https://github.com/openebs/lvm-localpv/issues/353 error I can't be sure about that

mhkarimi1383 committed:
:technologist: Add `.envrc` to automatically load nix-shell

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

dsharma-dc commented on coverage.txt:
If I'm not mistaken, this is an autogenerated file and doesn't need to be checked-in.

mhkarimi1383 replied:
Yes, It should be in `.gitignore` :)
I have pushed that file by mistake

mhkarimi1383 committed:
:sparkles: Isolating tmpdir and gopath in nix-shell

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 commented:
Almost complete,
I have isolated nix-shell's gopath and tempdir, Since installing specific version of some packages is a bit messy

mhkarimi1383 committed:
:pencil2: Shell command order

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:pencil2: Set GOCACHE

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

tiagolobocastro commented on shell.nix:
```diff
@@ -5,8 +5,24 @@ in
+    export TMPDIR=$(pwd)/nix/.tmp
+    export PATH=$GOPATH/bin:$PATH
+    mkdir -p "$TMPDIR"
+    go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.4.0
```
I think this is now the same the Makefile is installing? Do we need to do it here?

mhkarimi1383 replied:
But setting env variables was required, so I can run `make bootstrap` instead of `go install`

tiagolobocastro replied:
Yeah that's fine, just wondering about specifically the install to avoid having to modify this file once we update the Makefile

tiagolobocastro replied:
That said we also have the specific go version here, so we might need to change anyway :)

mhkarimi1383 replied:
Yeah, Also if we upgrade nixpkgs specified go version will not be available

mhkarimi1383 replied:
I would recommend changing the entire CI system to use nix-shell just like lint action

With that we will have a single source of truth :)

mhkarimi1383 replied:
Switched to `make bootstrap`, but getting

```
go: github.com/axw/gocov/gocov@latest (in github.com/axw/gocov@v1.2.1): go.mod:3: invalid go version '1.22.0': must match format 1.23
go.mod:5: unknown directive: toolchain
```

in the logs

but tools are installed correctly as far as I can tell

mhkarimi1383 replied:
Also getting

```
  go: gopkg.in/matm/v1/gocov-html@latest: unrecognized import path "gopkg.in": parse https://gopkg.in/?go-get=1: no go-import meta tags ()
```


Fixed in https://github.com/openebs/lvm-localpv/pull/354

mhkarimi1383 committed:
:pencil2: Use `make bootstrap` instead of `go install`

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:see_no_evil: Add coverage file to gitignore

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

mhkarimi1383 committed:
:green_heart: Fix HOME env var on lint action

Signed-off-by: Muhammed Hussein Karimi <info@karimi.dev>

tiagolobocastro commented on shell.nix:
```diff
@@ -5,8 +5,25 @@ in
-}
\ No newline at end of file
+  shellHook = ''
+    export HOME=${builtins.getEnv "HOME"}
```
why is this needed?

mhkarimi1383 replied:
In GitHub actions it's required to set HOME specifically, it will fail if you don't set that (in both action file and shell.nix)

mhkarimi1383 replied:
You can checkout previous commits action runs

tiagolobocastro replied:
Interesting, was not aware of this :+1:
